src/transformers/generation/utils.py

```diff
@@ -406,7 +406,10 @@ def prepare_inputs_for_generation(
             ):
                 input_ids = input_ids[:, -cache_position.shape[0] :]
             elif input_ids.shape[1] != cache_position.shape[0]:  # Default case (the "else", a no op, is Exception 2)
-                input_ids = input_ids[:, cache_position]
+                if input_ids.device.type == "hpu":
+                    input_ids = input_ids.index_select(1, cache_position)
+                else:
+                    input_ids = input_ids[:, cache_position]
```

Comment:
this is ugly and we can remove it if it's problematic.

Reply:
why not use index select for everyone?

Reply:
It seems this fix is not needed on the new version of SynapseAI 1.20.0 + Pytorch 2.6 + Gaudi2, so we'll probably target these two as minimum requirement https://huggingface.slack.com/archives/C03264192KV/p1741272959824279?thread_ts=1739977452.256159&cid=C03264192KV.